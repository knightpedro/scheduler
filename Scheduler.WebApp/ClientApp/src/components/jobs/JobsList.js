import React from "react";
import { List } from "semantic-ui-react";
import { useSelector } from "react-redux";
import { jobsSelectors } from "../../ducks/jobs";
import { Empty } from "../common";

const JobsList = ({ activeId, handleClick, filter }) => {
  const jobs = useSelector((state) =>
    jobsSelectors.selectFiltered(state, filter)
  );
  if (!jobs || jobs.length === 0) return <Empty message="No jobs found" />;

  return (
    <List selection>
      {jobs.map((j) => (
        <List.Item
          key={j.id}
          onClick={() => handleClick(j.id)}
          active={j.id === activeId}
        >
          <List.Icon
            verticalAlign="middle"
            name={j.isComplete ? "check" : "refresh"}
            color={j.isComplete ? "green" : "blue"}
          />
          <List.Content>
            <List.Header>{j.jobNumber}</List.Header>
            <List.Description>{j.description}</List.Description>
          </List.Content>
        </List.Item>
      ))}
    </List>
  );
};

export default JobsList;
