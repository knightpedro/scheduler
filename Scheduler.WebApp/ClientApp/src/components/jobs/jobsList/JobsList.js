import React, { useMemo } from "react";
import moment from "moment";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faSyncAlt } from "@fortawesome/free-solid-svg-icons";
import Routes from "../../../routes";
import { generatePath } from "react-router-dom";
import {
  BasicTable,
  SearchColumnFilter,
  SelectColumnFilter
} from "../../common/tables";

const DATE_FORMAT = "DD/MM/YYYY";

const status = {
  COMPLETE: "Complete",
  IN_PROGRESS: "In Progress"
};

const JobsList = ({ jobs }) => {
  const createPath = id => generatePath(Routes.jobs.DETAIL, { id });

  const columns = useMemo(
    () => [
      {
        Header: "Job Number",
        accessor: "jobNumber",
        Cell: ({ row }) => (
          <Link to={createPath(row.original.id)}>{row.values.jobNumber}</Link>
        ),
        Filter: SearchColumnFilter
      },
      {
        Header: "Description",
        accessor: "description",
        Filter: SearchColumnFilter
      },
      {
        Header: "Location",
        accessor: "location",
        Filter: SearchColumnFilter
      },
      {
        Header: "Complete",
        accessor: "status",
        className: "text-center",
        Cell: ({ cell: { value } }) => (
          <FontAwesomeIcon
            color={value === status.COMPLETE ? "green" : "blue"}
            icon={value === status.COMPLETE ? faCheck : faSyncAlt}
          />
        ),
        Filter: SelectColumnFilter,
        filter: "exactText"
      },
      {
        Header: "Date Received",
        accessor: "dateReceived",
        className: "text-center",
        Cell: ({ cell: { value } }) => value.format(DATE_FORMAT),
        disableFilters: true,
        sortType: "momentSort"
      }
    ],
    []
  );

  const data = useMemo(
    () =>
      jobs.map(job => ({
        id: job.id,
        jobNumber: job.jobNumber,
        description: job.description,
        location: job.location,
        status: job.isComplete ? status.COMPLETE : status.IN_PROGRESS,
        dateReceived: moment(job.dateReceived)
      })),
    [jobs]
  );

  const initialTableState = useMemo(
    () => ({
      sortBy: [
        {
          id: "dateReceived",
          desc: true
        }
      ]
    }),
    []
  );

  return (
    <BasicTable
      columns={columns}
      data={data}
      initialState={initialTableState}
    />
  );
};

export default JobsList;
