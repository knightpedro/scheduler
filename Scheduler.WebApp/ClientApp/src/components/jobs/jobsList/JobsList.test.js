import React from "react";
import { shallow } from "enzyme";
import JobsList from "./JobsList";
import moment from "moment";

const jobStub = {
    id: 1,
    jobNumber: "J123",
    description: "Test Job",
    location: "Test Room",
    isComplete: false,
};

jest.mock("moment");

describe("JobsList", () => {
    it("renders correctly", () => {
        const wrapper = shallow(<JobsList jobs={[jobStub]} />);
        expect(wrapper).toMatchSnapshot();
    });
});
