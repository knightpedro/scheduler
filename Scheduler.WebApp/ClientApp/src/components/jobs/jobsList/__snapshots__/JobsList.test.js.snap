// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsList renders correctly 1`] = `
<BasicTable
  columns={
    Array [
      Object {
        "Cell": [Function],
        "Filter": [Function],
        "Header": "Job Number",
        "accessor": "jobNumber",
      },
      Object {
        "Filter": [Function],
        "Header": "Description",
        "accessor": "description",
      },
      Object {
        "Filter": [Function],
        "Header": "Location",
        "accessor": "location",
      },
      Object {
        "Cell": [Function],
        "Filter": [Function],
        "Header": "Complete",
        "accessor": "status",
        "className": "text-center",
        "filter": "exactText",
      },
      Object {
        "Cell": [Function],
        "Header": "Date Received",
        "accessor": "dateReceived",
        "className": "text-center",
        "disableFilters": true,
        "sortType": "momentSort",
      },
    ]
  }
  data={
    Array [
      Object {
        "dateReceived": undefined,
        "description": "Test Job",
        "id": 1,
        "jobNumber": "J123",
        "location": "Test Room",
        "status": "In Progress",
      },
    ]
  }
  initialState={
    Object {
      "sortBy": Array [
        Object {
          "desc": true,
          "id": "dateReceived",
        },
      ],
    }
  }
/>
`;
